docs/source/references/schema.json

```diff
@@ -1,5404 +1,3435 @@
+        "type": "PeerEventFilter",
+        "discriminant": 1
+      }
+    ]
+  },
+  "FilterOpt<PeerFilter>": {
+    "Enum": [
+      {
+        "name": "AcceptAll",
+        "discriminant": 0
+      },
+      {
+        "name": "BySome",
+        "type": "PeerFilter",
+        "discriminant": 1
+      }
+    ]
+  },
+  "FilterOpt<RoleEventFilter>": {
+    "Enum": [
+      {
+        "name": "AcceptAll",
+        "discriminant": 0
+      },
+      {
+        "name": "BySome",
+        "type": "RoleEventFilter",
+        "discriminant": 1
+      }
+    ]
+  },
+  "FilterOpt<RoleFilter>": {
+    "Enum": [
+      {
+        "name": "AcceptAll",
+        "discriminant": 0
+      },
+      {
+        "name": "BySome",
+        "type": "RoleFilter",
+        "discriminant": 1
+      }
+    ]
+  },
+  "FilterOpt<TriggerEventFilter>": {
+    "Enum": [
+      {
+        "name": "AcceptAll",
+        "discriminant": 0
+      },
+      {
+        "name": "BySome",
+        "type": "TriggerEventFilter",
+        "discriminant": 1
+      }
+    ]
   },
-  "Option<iroha_data_model::domain::IpfsPath>": {
-    "Option": "iroha_data_model::domain::IpfsPath"
+  "FilterOpt<TriggerFilter>": {
+    "Enum": [
+      {
+        "name": "AcceptAll",
+        "discriminant": 0
+      },
+      {
+        "name": "BySome",
+        "type": "TriggerFilter",
+        "discriminant": 1
+      }
+    ]
   },
-  "Option<iroha_data_model::events::pipeline::EntityKind>": {
-    "Option": "iroha_data_model::events::pipeline::EntityKind"
+  "FindAccountById": {
+    "Struct": {
+      "id": "EvaluatesTo<AccountId>"
+    }
   },
-  "Option<iroha_data_model::events::pipeline::StatusKind>": {
-    "Option": "iroha_data_model::events::pipeline::StatusKind"
+  "FindAccountKeyValueByIdAndKey": {
+    "Struct": {
+      "id": "EvaluatesTo<AccountId>",
+      "key": "EvaluatesTo<Name>"
+    }
   },
-  "Option<iroha_data_model::events::time::Interval>": {
-    "Option": "iroha_data_model::events::time::Interval"
+  "FindAccountsByDomainId": {
+    "Struct": {
+      "domain_id": "EvaluatesTo<DomainId>"
+    }
   },
-  "Option<iroha_data_model::isi::Instruction>": {
-    "Option": "iroha_data_model::isi::Instruction"
+  "FindAccountsByName": {
+    "Struct": {
+      "name": "EvaluatesTo<Name>"
+    }
   },
-  "Option<iroha_data_model::name::Name>": {
-    "Option": "iroha_data_model::name::Name"
+  "FindAccountsWithAsset": {
+    "Struct": {
+      "asset_definition_id": "EvaluatesTo<AssetDefinitionId>"
+    }
   },
-  "Option<u32>": {
-    "Option": "u32"
+  "FindAllAccounts": null,
+  "FindAllActiveTriggerIds": null,
+  "FindAllAssets": null,
+  "FindAllAssetsDefinitions": null,
+  "FindAllBlockHeaders": null,
+  "FindAllBlocks": null,
+  "FindAllDomains": null,
+  "FindAllParameters": null,
+  "FindAllPeers": null,
+  "FindAllPermissionTokenDefinitions": null,
+  "FindAllRoleIds": null,
+  "FindAllRoles": null,
+  "FindAllTransactions": null,
+  "FindAssetById": {
```

Comment:
But there is only one reference (to each):

```rust
enum QueryBox {
  FindAllAccounts(FindAllAccounts),
  // ...
}
```

Isn't it possible to remove an associated unit struct at all?

```rust
enum QueryBox {
  FindAllAccounts,
  // ...
}
```

Reply:
This is how it looks in schema:
```json
  "QueryBox": {
    "Enum": [
      {
        "name": "FindAllAccounts",
        "type": "FindAllAccounts"
      }
      ...
    ]
}
```

Look at `"type"`. There `QueryBox` references a type `FindAllAccounts`. We don't know what it is so we go and look up it's definition in the schema and we find this:
```json
  "FindAllAccounts": null,
```

This is the usual schema resolution process. I suppose that we can flatten it like this or something similar:
```rs
```json
  "QueryBox": {
    "Enum": [
      {
        "name": "FindAllAccounts",
        "type": null
      }
      ...
    ]
}
```

but then we can do that for any type, it's not that `()` is special

Reply:
Isn't it possible to remove these unit structures completely from the code base? Do they make any sense in Rust code?

Reply:
> Isn't it possible to remove these unit structures completely from the code base? Do they make any sense in Rust code?

Yes, they make sense in the rust code, and no they cannot be removed from the codebase. They are not just part of the `QueryBox` they can be used on their own. `QueryBox` is only a convenience serialization format. It just so happens that they don't have any fields inside.

Maybe you would like to serialize them like this:
```json
  "FindAllAccounts": {},
```

but this might not be conveying the appropriate semantics. I've used `serde::Serializer::serialize_unit_struct` precisely because this is a unit struct

Reply:
> Yes, they make sense in the rust code, and no they cannot be removed from the codebase.

Okay then. Is it possible to remain them in Rust, but make them invisible from the perspective in schema? I think it makes sense because there is no point to have these structs in the schema.

Reply:
The point of the schema is to show you how it's done in the reference implementation (i.e. Rust) so that a dev can reflect this API in a specific sdk. We should strive to have a uniform developing experience of Iroha across different programming languages. If in schema you see:
```json
"QueryBox": {
  "Enum": [
  {
    "name": "FindAllAccounts",
    "type": "FindAllAccounts"
  }
  ...
  ]
}
"FindAllAccounts": null,
```

It means that in rust it looks and is used like:
```rs
pub struct FindAllAccounts;
pub enum QueryBox {
    FindAllAccounts(FindAllAccounts),
}

let query = FindAllAccounts;
let query = FindAllAccounts {}; // both are valid
let boxed_query = QueryBox::from(query); // if you want to box the query
```

whereas in Java it might look close to this:
```java
public interface Query {}

public class FindAllAccounts implements Query {
    // Schema says that this type should exist and be empty
}

public class QueryBox {
    Query query;

    public<Q extends Query> QueryBox from(Q query) {
        this.query = query;
    }
}

FindAllAccounts query = FindAllAccounts();
QueryBox boxed_query = QueryBox::from(query);
```

I don't know whether or not it will be possible for an sdk API to reflect Rust API or will there be some limitations but at least they will know by looking at schema how it is supposed to look like.

Reply:
> The point of the schema is to show you how it's done in the reference implementation (i.e. Rust)

I am completely disagree here. I think that the point of the schema is to be completely agnostic to any implementation and to reflect pure Iroha data model namespace and how it is represented in SCALE. Ideally, even Rust side should depend on the schema, not the opposite. It shouldn't have Rust implementation details, it should only have SCALE spec details. The schema should be the source of truth which is established first, and which is then implemented on each platform. 

That's why I think these structs are redundand - if the schema was the source of truth, it wouldn't have this extra noise.

Reply:
Keeping the schema as the source of truth is also good in the terms of documentations and availability of API to the users, IMO.

Reply:
> I am completely disagree here. I think that the point of the schema is to be completely agnostic to any implementation and to reflect pure Iroha data model namespace and how it is represented in SCALE. Ideally, even Rust side should depend on the schema, not the opposite. It shouldn't have Rust implementation details, it should only have SCALE spec details. The schema should be the source of truth which is established first, and which is then implemented on each platform.

ok, I see where you're coming from and I mostly agree with that view point. The question then is: should we allow for empty type definitions in the schema. Can we say like this: if in the schema there is a type definition that is empty like in `"FindAllAccounts": null`, an sdk developer is free to not implement the empty type, but if he chooses to implement the empty type, type should be named `FindAllAccounts`

Reply:
I would also encourage the dev to implement the empty type, but I'm not sdk dev so I don't have enough broad perspective to do so

Reply:
I'm ok with changing the representation of an empty type as long as we agree to keep them. Some of possible variations are:
```json
  "FindAllAccounts": {
    "Tuple": []
    }
```
```json
  "FindAllAccounts": {
    "Struct": {}
    }
```

but serde decides for me that the best format is `FindAllAccounts: null` because this format doesn't assume anything about the structure of an empty type (it can be struct, enum, tuple, array, ...)

Reply:
> Can we say like this: if in the schema there is a type definition that is empty like in "FindAllAccounts": null, an sdk developer is free to not implement the empty type, but if he chooses to implement the empty type, type should be named FindAllAccounts

Now it works this way, at least in JS SDK.

> I would also encourage the dev to implement the empty type, but I'm not sdk dev so I don't have enough broad enough perspective to give such advices

There is a type to handle unit structs, at least in JS SDK.

---

My point is to ask the question from the perspective of schema, without knowing about Rust impl: does it make sense to have `FindAllAccounts`, or in general, `QueryBox::FindAllAccounts(())`, instead of having just `QueryBox::FindAllAccounts`? I would answer "No, it doesn't, it's just a noise and the second variant is clearer."

Reply:
opened #3359